src/librustc_mir/borrow_check/nll/explain_borrow/mod.rs

```diff
@@ -19,69 +19,75 @@ use util::liveness::{self, DefUse, LivenessMode};
 
 impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
     pub(in borrow_check) fn explain_why_borrow_contains_point(
-        &self,
+        &mut self,
         context: Context,
         borrow: &BorrowData<'tcx>,
         err: &mut DiagnosticBuilder<'_>,
     ) {
         if let Some(regioncx) = &self.nonlexical_regioncx {
-            if let Some(cause) = regioncx.why_region_contains_point(borrow.region, context.loc) {
-                let mir = self.mir;
-
-                match *cause.root_cause() {
-                    Cause::LiveVar(local, location) => {
-                        match find_regular_use(&mir, regioncx, borrow, location, local) {
-                            Some(p) => {
-                                err.span_label(
-                                    mir.source_info(p).span,
-                                    format!("borrow later used here"),
-                                );
-                            }
+            let mir = self.mir;
 
-                            None => {
-                                span_bug!(
-                                    mir.source_info(context.loc).span,
-                                    "Cause should end in a LiveVar"
-                                );
+            if self.nonlexical_cause_info.is_none() {
+                self.nonlexical_cause_info = Some(regioncx.compute_causal_info(mir));
+            }
+
+            if let Some(cause_info) = &self.nonlexical_cause_info {
```

Comment:
can you rewrite this as:

```rust
let cause_info = self.nonlexical_cause_info.as_ref().unwrap();
```

after all, it had *better* be `Some`, since we just assigned it up there!